### PR TITLE
Smart-lock Android 12

### PR DESCRIPTION
**Pull request by jonesc-oci**
Smart-lock updates for Android 12, OpenDDS 3.21.

### CODE REVIEW

File: android/SmartLock/app/src/main/java/org/opendds/smartlock/OpenDdsBridge.java [hunk 1]
```diff
@@ -32,39 +31,30 @@
...
+    public OpenDdsBridge(MainActivity activity) {
         this.activity = activity;
+        this.context = activity.getApplicationContext();
     }
 
-    // need a persistent reference to datawriter to avoid GC
-    private static DataWriter dw = null;
-
-    public boolean updateLockState (SmartLockStatus lockState)
+    public void updateLockState (SmartLockStatus lockState)
```
Comment (mitza-oci) [inline android/SmartLock/app/src/main/java/org/opendds/smartlock/OpenDdsBridge.java:57]:
```suggestion
    public void updateLockState(SmartLockStatus lockState)
```

There's a bunch of these (mostly in code that's not changed in this PR)

File: android/SmartLock/gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.4-all.zip
```
Comment (mitza-oci) [inline android/SmartLock/gradle/wrapper/gradle-wrapper.properties:6]: Is this `\:` "escaping" normal in the properties file?  Looks odd to me, but I don't usually work with these

File: src/SmartLock.cpp [hunk 3]
```diff
@@ -30,32 +30,40 @@ std::ostream& operator<< (std::ostream& lhs, const SmartLock::lock_t& rhs) {
...
+void pi_shutdown() {
+  gpioTerminate();
 }
 
 void pi_init() {
-  wiringPiSetup();
-  pinMode(LOCKED_LIGHT, OUTPUT);
-  pinMode(UNLOCKED_LIGHT, OUTPUT);
+  if (gpioInitialise() < 0) {
+    ACE_ERROR((LM_ERROR, "ERROR: pigpio initialisation failed\n"));
```
Comment (mitza-oci) [inline src/SmartLock.cpp:61]:
```suggestion
    ACE_ERROR((LM_ERROR, "ERROR: pigpio initialization failed\n"));
```
Even if function names are en-GB, error message should be en-US for consistency (or log the function name as-is)

File: dockerfiles/pi-opendds-cross/build.sh [hunk 4]
```diff
@@ -5,55 +5,61 @@ OPENDDS_GIT_REPO=https://github.com/objectcomputing/OpenDDS
 OPENDDS_MAKE_JOBS=4
 OPENDDS_PREFIX=${BUILD_DIR}/pi-opendds
 
-wget -nc http://download.objectcomputing.com/OpenDDS/resources/gcc-8.2.0-rpi.tar.bz2
-tar xjf gcc-8.2.0-rpi.tar.bz2
+wget -nc -O gcc-arm-10.2-2020.11-x86_64-arm-none-linux-gnueabihf.tar.xz \
```
Comment (mitza-oci) [inline dockerfiles/pi-opendds-cross/build.sh:8]: Is there a way to reduce duplication between Dockerfile `RUN` script commands and the `build.sh`?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: dockerfiles/pi-opendds-cross/build.sh (hunk 4 above)
Comment (jonesc-oci) [reply dockerfiles/pi-opendds-cross/build.sh:8]: Yes. The docker image can include and run build.sh.

File: dockerfiles/android-opendds-cross/Dockerfile [hunk 5]
```diff
@@ -28,34 +45,44 @@ USER droid
...
+        https://dlcdn.apache.org//xerces/c/3/sources/xerces-c-3.2.3.tar.gz && \
+        tar xvzf xerces-c-3.2.3.tar.gz
 
 RUN printf "\
 set(CMAKE_SYSTEM_NAME Linux)\n \
 set(CMAKE_SYSTEM_PROCESSOR arm)\n \
-set(CMAKE_C_COMPILER aarch64-linux-android28-clang)\n \
-set(CMAKE_CXX_COMPILER aarch64-linux-android28-clang++)\n \
+set(CMAKE_C_COMPILER ${ANDROID_CC})\n \
+set(CMAKE_CXX_COMPILER ${ANDROID_CXX})\n \
```
[the remark this answers is not in the document]
Comment (mitza-oci) [reply dockerfiles/android-opendds-cross/Dockerfile:85]: The docker build process could get a newer CMake by downloading from cmake.org or using https://apt.kitware.com/ packages (though it looks like those are only Ubuntu and not Debian)

[the remark this answers is not in the document]
Comment (jonesc-oci) [reply dockerfiles/android-opendds-cross/Dockerfile]: Changed so that image downloads CMake

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**